File: src/SlimShader/Chunks/Xsgn/Name.cs
using SlimShader.Chunks.Shex;
using SlimShader.Util;

namespace SlimShader.Chunks.Xsgn
{
	/// <summary>
	/// Values that identify shader parameters that use system-value semantics.
	/// Based on D3D_NAME. This is similar to, but different from, <see cref="SystemValueName"/>
	/// </summary>
	public enum Name
	{
		[Description("NONE")]
		Undefined = 0,

		[Description("POS")]
		Position = 1,

		ClipDistance = 2,
		CullDistance = 3,

		[Description("RTINDEX")]
		RenderTargetArrayIndex = 4,

		ViewportArrayIndex = 5,

		[Description("VERTID")]
		VertexID = 6,

		[Description("PRIMID")]
		PrimitiveID = 7,

		InstanceID = 8,
		IsFrontFace = 9,
		SampleIndex = 10,

		[Description("QUADEDGE")]
		FinalQuadEdgeTessFactor = 11,

		[Description("QUADINT")]
		FinalQuadInsideTessFactor = 12,

		[Description("TRIEDGE")]
		FinalTriEdgeTessFactor = 13,

		[Description("TRIINT")]
		FinalTriInsideTessFactor = 14,

		[Description("LINEDET")]
		FinalLineDetailTessFactor = 15,

		[Description("LINEDEN")]
		FinalLineDensityTessFactor = 16,

		[Description("TARGET")]
		Target = 64,

		[Description("DEPTH")]
		Depth = 65,

		Coverage = 66,

		[Description("DEPTHGE")]
		DepthGreaterEqual = 67,

		[Description("DEPTHLE")]
		DepthLessEqual = 68
	}
}

File: src/cpp/Source/Chunks/Xsgn/Name.cpp
#include "PCH.h"
#include "Name.h"

using namespace std;
using namespace SlimShader;

string SlimShader::ToString(Name value)
{
	switch (value)
	{
	case Name::Undefined :
		return "NONE";
	case Name::Position :
		return "POS";
	case Name::RenderTargetArrayIndex :
		return "RTINDEX";
	case Name::VertexID :
		return "VERTID";
	case Name::PrimitiveID :
		return "PRIMID";
	case Name::FinalQuadEdgeTessFactor :
		return "QUADEDGE";
	case Name::FinalQuadInsideTessFactor :
		return "QUADINT";
	case Name::FinalTriEdgeTessFactor :
		return "TRIEDGE";
	case Name::FinalTriInsideTessFactor :
		return "TRIINT";
	case Name::FinalLineDetailTessFactor :
		return "LINEDET";
	case Name::FinalLineDensityTessFactor :
		return "LINEDEN";
	case Name::Target :
		return "TARGET";
	case Name::Depth :
		return "DEPTH";
	case Name::DepthGreaterEqual :
		return "DEPTHGE";
	case Name::DepthLessEqual :
		return "DEPTHLE";
	default :
		throw runtime_error("Unsupported value: " + to_string((int) value));
	}
}